# Release 4.1.0

golluroSICKAG opened the pull request:
# Version 4.1.0

## New features
- Provide version of module via 'OnNewStatusModuleVersion'
- Function 'saveAllModuleConfigs' to trigger multiple / all running CSK modules to save their persistent data
- Function to reload all apps ('reloadApps')
- Function to reboot device ('rebootDevice')
- Check if features of module can be used on device and provide this via 'OnNewStatusModuleIsActive' event / 'getStatusModuleActive' function
- Function 'resetAllModules' and event 'OnResetAllModules' so that other modules can react by loading their default setup
- Function 'removeData' to remove all data of current DataSet
- Extra thread to trigger other modules to load their specific parameters if event 'OnNewDataToLoad' was notified

## Improvements
- New parameter for 'loadContent' to controle if other modules directly should react to new parameters
- New UI design available (selectable via 'setUIStyle')
- Added UI icon and browser tab information

## Following tasks were checked

- [x] Module is coded and structured according to the "Developing guideline for modules" described within the CSK documentation
- [x] All functions/events/parameters are documented within the manifest documentation
- [x] The manifest description of the main CROWN includes main information about the purpose of the module and how to use it in general
- [x] API docu based on the manifest was created and stored within the "docu"-folder of the repository
- [x] Internal LUA code documentation exists for variables and non served functions
- [x] All relevant infos are logged via the SharedLogger 'ModuleLogger'
- [x] Module supports persistent data feature based on 'CSK_Module_PersistentData'
- [x] Module supports user management based on 'CSK_Module_UserManagement'
- [x] No open "ToDos" within the code or at least clearly explained comments why they exist...
- [x] "Version" key in app manifest was updated following semantic versioning (and use '0.x.y' for test / experimental modules which are not yet ready to be officially released)
- [x] Meaningful IDs used for UI elements
- [x] Module was tested on an AppSpace device (at least on SICK AppEngine) with no error message
- [x] README.md is up to date (incl. info of device + firmware the module was tested with)
- [x] CHANGELOG.md is up to date

golluroSICKAG committed:
# Release 4.1.0

## New features
- Provide version of module via 'OnNewStatusModuleVersion'
- Function 'saveAllModuleConfigs' to trigger multiple / all running CSK modules to save their persistent data
- Function to reload all apps ('reloadApps')
- Function to reboot device ('rebootDevice')
- Check if features of module can be used on device and provide this via 'OnNewStatusModuleIsActive' event / 'getStatusModuleActive' function
- Function 'resetAllModules' and event 'OnResetAllModules' so that other modules can react by loading their default setup
- Function 'removeData' to remove all data of current DataSet
- Extra thread to trigger other modules to load their specific parameters if event 'OnNewDataToLoad' was notified

## Improvements
- New parameter for 'loadContent' to controle if other modules directly should react to new parameters
- New UI design available (selectable via 'setUIStyle')
- Added UI icon and browser tab information

golluroSICKAG committed:
- Minor docu in code

golluroSICKAG committed:
- Function to reboot device not supported by SAE and Emulator

golluroSICKAG committed:
- Update of UI settings

kogliinSICKAG commented on CSK_Module_PersistentData/scripts/Configuration/PersistentData/PersistentData_Controller.lua:
```diff
@@ -234,13 +260,97 @@ end
+local function rebootDevice()
+  local typeName = Engine.getTypeName()
+  if typeName == 'AppStudioEmulator' or typeName == 'SICK AppEngine' then
+    _G.logger:warning(nameOfModule .. ': Function to reboot not supported by device!')
```
Would be good to have this message not only within the Logger module but also as a callout within the GUI where the button has been clicked.

golluroSICKAG replied:
Done

kogliinSICKAG commented on README.md:
```diff
@@ -13,10 +13,13 @@ For further information regarding the internally used functions / events, check
+|SICK AppEngine|v1.7.0|v4.1.0|
 |SICK AppEngine|v1.5.0|v4.0.1|
 |InspectorP631|v4.2.1|v4.0.1|
+|SIM1012|v2.4.2|v4.1.0|
```
SIM1012 latest firmware is v2.4.1

kogliinSICKAG commented on CSK_Module_PersistentData/pages/pages/CSK_Module_PersistentData/CSK_Module_PersistentData.html:
```diff
@@ -1,137 +1,407 @@
-		</stacked-view>
-	</layout-column>
-</layout-row>
+<layout-row id="RowLayout1" class="myCustomBackground_CSK_Module_PersistentData">
```
Would be nice to have a button, to delete a modules complete parameter set from the current parameters file or did I not recognize it?

golluroSICKAG replied:
I added some more documentation regarding how to clear parametets.
To delete files, please make use of the CSK_Module_FileManager

golluroSICKAG committed:
- Fixed issues of review

- Added function 'getCurrentParameterInfo' to get info what parameter file is currently in use

golluroSICKAG commented:
Thx for the review!

golluroSICKAG committed:
- Improved procedure to check if modules send their parameters if it was triggered that all modules should save them

golluroSICKAG committed:
- Within saveAll, check if module is active. Otherwise do not save parameters

golluroSICKAG committed:
- Removing navigation bar entries
